app/src/processing/app/Editor.java

```diff
@@ -1265,43 +1260,58 @@ public void actionPerformed(ActionEvent e) {
     return menu;
   }
 
+  /**
+   * Wrapper around RSyntaxTextArea's RecordableTextActions. Normally,
+   * these actions use the event source, or the most recently focused
+   * component to find the text area to work on, but this does not
+   * always work. This wrapper makes them always work on the currently
+   * selected tab instead.
+   */
+  class RstaActionWrapper extends ActionWrapper {
+    RstaActionWrapper(RecordableTextAction action) {
+      super(action);
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+      ((RecordableTextAction) getWrappedAction()).actionPerformedImpl(e, getCurrentTab().getTextArea());
+    }
+  }
 
   private JMenu buildEditMenu() {
     JMenu menu = new JMenu(tr("Edit"));
     menu.setName("menuEdit");
     menu.setMnemonic(KeyEvent.VK_E);
 
-    undoItem = newJMenuItem(tr("Undo"), 'Z');
+    // Create a dummy RTextArea, to force it to create the static list
+    // of actions accessible through RTextArea.getAction().
+    new RTextArea();
+
+    RecordableTextAction undoAction = RTextArea.getAction(RTextArea.UNDO_ACTION);
+    JMenuItem undoItem = new JMenuItem(new RstaActionWrapper(undoAction));
```

Comment:
The reason to use RSTA's actions, is that they are automatically updated to become enabled/disabled when the undo state changes, and to automatically show "Undo" or "Can't undo". If we do that manually, as is done currently, we need access to the UndoManager which needs a patch (at least for the `getUndoPresentationName()`, IIRC RSTA does export a `canUndo()` or similar). Also, we need to keep track of changes within the document with change listeners and other stuff, which is a bit messy. Having said that, this approach isn't 100% elegant either, so I can understand your being hesitant.